xapian-applications/omega/date.cc

```diff
@@ -154,7 +163,14 @@ date_range_filter(const string & date_start, const string & date_end,
 	    t.tm_hour = 12;
 	    t.tm_min = t.tm_sec = 0;
 	    t.tm_isdst = -1;
-	    time_t then = mktime(&t) - secs;
+	    time_t then;
+	    // if secs is equal to time_t min value,
+	    // then = mktime + time_t min val hence which would do addition
+	    // so checking for that overflow.
+	    if (secs < (mktime(&t) - numeric_limits<time_t>::max()))
+		then = numeric_limits<time_t>::max();
+	    else
+		then = mktime(&t) - secs;
```

Comment:
Calling `mktime()` twice is probably better avoided, as it isn't a completely trivial calculation it's doing, and the compiler is unlikely to be able to tell that it can actually safely eliminate the second call.

I'm not convinced the underflow check here is correct - shouldn't we be comparing against the `min()` not `max()` (at least if secs is positive)?

Underflow happens when (comparison done in the full integer range):

```
mktime(&t) - secs < MIN
```

Which is the same as this condition, only this check won't underflow or overflow if `secs >= 0`):

```
mktime(&t) < MIN + secs
```